File: src/weibreg.c
#include <R.h>
#include <R_ext/Applic.h>
#include "weibreg.h"


static double we_fun(int n, double *beta, void *vex){

    int ord, ipfixed, i;
    Exts *ex;
    double f, ftot, *w_beta;
    int nn, start, mb, w_n;
    double *dummy = NULL;
    
    ex = vex;
    mb = *(ex->mb);
    w_n = mb + 2; 
    w_beta = Calloc(mb + 2, double); 

    ord = 0;
    ipfixed = 0;
    ftot = 0.0;
    
    for (i = 0; i < mb; i++) w_beta[i] = beta[i];
    for (i = 0; i < *(ex->ns); i++){
	start = ex->nstra[i];
	nn = ex->nstra[i+1] - start;
	w_beta[mb] = beta[mb + 2 * i];
	w_beta[mb + 1] = beta[mb + 2 * i + 1];
    
	F77_CALL(wfunc)(&ord, &ipfixed, ex->pfix, &w_n, ex->mb, w_beta,
			&nn, ex->z + mb * start, 
			ex->time0 + start, ex->time + start, 
			ex->ind + start, 
			ex->offset + start,
			&f, dummy, dummy, ex->iok);
	ftot += f;
    }
    Free(w_beta);
    return(ftot);
}
	
static void gwe_fun(int n, double *beta, double *dloglik, void *vex){

    int ord, ipfixed, i, j;
    double *fp, *w_beta;
    Exts *ex;
    int start, nn, mb, w_n;
    double f;
    double *dummy = NULL;

    ex = vex;

    mb = *(ex->mb);
    w_n = mb + 2;

    fp = Calloc(mb + 2, double);
    w_beta = Calloc(mb + 2, double);

    ord = 1;
    ipfixed = 0;


    for (j = 0; j < n; j++) dloglik[j] = 0.0;
    for (i = 0; i < mb; i++) w_beta[i] = beta[i];
    for (i = 0; i < *(ex->ns); i++){
	start = ex->nstra[i];
	nn = ex->nstra[i+1] - ex->nstra[i];
	w_beta[mb] = beta[mb + 2 * i];
	w_beta[mb + 1] = beta[mb + 2 * i + 1];
	F77_CALL(wfunc)(&ord, &ipfixed, ex->pfix, &w_n, ex->mb, w_beta,
			&nn, ex->z + *(ex->mb) * start,
			ex->time0 + start, ex->time + start,
			ex->ind + start, ex->offset + start,
			&f, fp, dummy, ex->iok);
	for (j = 0; j < mb; j++) dloglik[j] += fp[j];
	dloglik[mb + 2*i] += fp[mb];
	dloglik[mb + 2*i + 1] += fp[mb + 1];
    }
    Free(fp);
    Free(w_beta);
}
    
void sw_fun(int *order,
	    int *bdim, int *mb, double *beta, 
	    int *nn, double *z, double *time0, double *time, int *ind, 
	    double *offset, int *ns, int *nstra,
      	    double *f, double *fp, double *fpp, int *iok){

/* This function deals _only_ with true Weibull,   */
/* i.e., 'shape' is estimated.                     */
/* Data _must_ be sorted according to stratum.     */
/* (if ns == 1, no sorting is of course necessary) */
/* 'nstra' contains the 'cumsum' strata sizes,     */ 
/* starting with 0, length (ns + 1).               */
 
    int  i, j, m;
    double w_f, *w_fp, *w_fpp, *w_beta;
    int start, w_nn, w_bdim;
    int sc_row, sh_row;

    int ipfixed = 0;   /* 'shape', AKA 'p' is NOT fixed */ 
    double pfix = 0.0;
 
/* If no strata, do it quickly and return: */
    if (*ns == 1){ /* No strata */
	F77_CALL(wfunc)(order, &ipfixed, &pfix, bdim, mb, beta,
			nn, z,
			time0, time,
			ind, offset,
			f, fp, fpp, iok);
	return;
    }

/* If strata, a little elaboration is necessary.       */
/* We must call 'wfunc' for each stratum and add up.   */
/* Remember, each stratum has its own scale and shape. */

    w_bdim = *mb + 2;

    w_fp = Calloc(w_bdim, double);
    w_fpp = Calloc(w_bdim * w_bdim, double);
    w_beta = Calloc(w_bdim, double);

    *f = 0.0;
    for (j = 0; j < *bdim; j++) fp[j] = 0.0;
    for (j = 0; j < (*bdim) * (*bdim); j++) fpp[j] = 0.0;

    for (j = 0; j < *mb; j++) w_beta[j] = beta[j];
    for (i = 0; i < *ns; i++){
	start = nstra[i];
	w_nn = nstra[i+1] - nstra[i];
	sc_row = *mb + 2 * i;
	sh_row = sc_row + 1;
	w_beta[*mb] = beta[sc_row];
	w_beta[*mb + 1] = beta[sh_row];
	F77_CALL(wfunc)(order, &ipfixed, &pfix, &w_bdim, mb, w_beta,
			&w_nn, z + (*mb) * start,
			time0 + start, time + start,
			ind + start, offset + start,
			&w_f, w_fp, w_fpp, iok);
	*f += w_f;
	for (j = 0; j < *mb; j++) fp[j] += w_fp[j];
	fp[*mb + 2*i] += w_fp[*mb];
	fp[*mb + 2*i + 1] += w_fp[*mb + 1];
/* And then the hessian, a little bit complicated: */
	for (j = 0; j < *mb; j++){
	    for (m = 0; m <= j; m++){
		fpp[m + j * (*bdim)] += w_fpp[m + j * w_bdim];
	    }
	}

	for (m = 0; m < *mb; m++){
	    fpp[m + sc_row * (*bdim)] += w_fpp[m + (*mb) * w_bdim];
	    fpp[m + sh_row * (*bdim)] += w_fpp[m + ((*mb) + 1) * w_bdim];
	}

	fpp[sc_row + sc_row * (*bdim)] += w_fpp[*mb + (*mb) * w_bdim];
	fpp[sc_row + sh_row * (*bdim)] += w_fpp[*mb + ((*mb) + 1) * w_bdim];
	fpp[sh_row + sh_row * (*bdim)] += 
	    w_fpp[*mb + 1 + ((*mb) + 1) * w_bdim];
	
    }

    /* Fill in "the other" half: */
    for (j = 0; j < *bdim; j++){
	for (m = (j + 1); m < *bdim; m++){
	    fpp[m + j * (*bdim)] = fpp[j + m * (*bdim)];
	}
    }

    Free(w_fp);
    Free(w_fpp);
    Free(w_beta);
}


	
void weibsup(int *iter, double *eps, int *printlevel,
	     int *ns, int *nstra, int *nn, int *ncov, int *bdim,
	     double *time0, double *time, int * ind,
	     double *covar, double *offset,
	     double *init, double *beta, double *lambda, double *lambda_sd,
	     double *shape, double *shape_sd,
	     double *loglik, double *dloglik, double *variance, double *sctest,
	     int *conver, int *fail){

    Exts *ex;
    int i, j;
    int iok;
    int maxiter;
    int trace;
    int *mask;
    int events;
    int nREPORT = 1;
    int fncount, grcount;
    double *xin;
    double Fmin;
    /*    int zero = 0;
    int two = 2;
    double pfix = 1.0;
    */
    ex = (Exts *)R_alloc(1, sizeof(Exts));
    mask = (int *)R_alloc(*bdim, sizeof(int));
    xin = (double *)R_alloc(*bdim, sizeof(double));

    maxiter = 1000;
    trace = *printlevel;

    iok = 0;

/* Fill in 'ex': */
    ex->ns = ns;
    ex->nstra = nstra;
    ex->pfix = shape;
    ex->mb = ncov;
    ex->nn = nn;
    ex->z = covar;
    ex->time0 = time0;
    ex->time = time;
    ex->ind = ind;
    ex->f = loglik + 1; /* Note: 'loglik' is a vector of length 2 */ 
    ex->fp = dloglik;
    ex->fpp = variance;
    ex->offset = offset;
    ex->iok = &iok;

/* Calculate simple start values for first call to 'vmmin': */
    for (i = 0; i < *ncov; i++) {
	xin[i] = init[i];
	beta[i] = init[i];
    }

    for (i = *ncov; i < *bdim; i++){
	xin[i] = 0.0;
	beta[i] = 0.0;
    }

    *lambda = 0.0;
    events = 0;
    for (i = 0; i < *nn; i++){
	*lambda += time[i] - time0[i];
	events += ind[i];
    }

    if (events <= 0) error("No events\n");
    if (*lambda <= 0.0) error("No (or negative) exposure time!\n");
    *lambda = (double)events / *lambda;
    for (i = 0; i < *ns; i++){
	j = *ncov + 2 * i; 
	xin[j] = log(*lambda);
	beta[j] = log(*lambda);
    }

/* Done with initial values; the same p and lambda in all strata. */

    /* Temporary hack for check: */
/*
    for (i = 0; i < *bdim; i++) xin[i] = 0.1;
    xin[*bdim - 2] = 0.0;
    Fmin = we_fun(*bdim, xin, ex);
    gwe_fun(*bdim, xin, dloglik, ex);
    F77_CALL(wfunc)(&two, &zero, &pfix, bdim, ncov, xin,
		    nn, covar, time0, time, ind, offset,
		    &Fmin, dloglik, variance, &iok);
    Rprintf("\n!!b = 0.1:\n");
    Rprintf("f = %f\n", Fmin);
    Rprintf("fp: ");
    for (i = 0; i < *bdim; i++){
	Rprintf("%f ", dloglik[i]);
    }

    Rprintf("\n\n fpp = \n");
    for (i = 0; i < *bdim; i++){
	for (j = 0; j < *bdim; j++){
	    Rprintf("%f ", variance[j + *bdim * i]);
	}
	Rprintf("\n");
    }
*/
/*
    trace = 0;
    nmmin(*bdim, xin, beta, &Fmin, we_fun, 
	  fail, *eps, *eps, ex, 
	  1.0, 0.5, 2.0, trace,
	  &fncount, maxiter);
    trace = *printlevel;

    loglik[1] = -Fmin;
    gwe_fun(*bdim, beta, dloglik, ex);
    Rprintf("\nEfter 'nmmin': loglik = %f\n", -Fmin);
    Rprintf(" beta och dloglik:\n");
    for (i = 0; i < *bdim; i++){
	Rprintf("%f, %f\n", beta[i], dloglik[i]);
    }
*/

/* 'Mask out' the regression coefficients: */
    for (i = 0; i < *ncov; i++){
	mask[i] = 0;
    }

    for (i = *ncov; i < *bdim; i++){
	mask[i] = 1;
    }

/* Estimate the 'null' model (only scale and shape): */
    vmmin(*bdim, beta, &Fmin,
	  we_fun, gwe_fun, maxiter, trace,  
	  mask, *eps, *eps, nREPORT,
	  ex, &fncount, &grcount, fail);

    if (trace)
	Rprintf("\nOnly scale and shape: loglik = %f\n", -Fmin);

    loglik[0] = -Fmin;

/* Now, 'mask in' the regression coefficients: */
    for (i = 0; i < *bdim; i++){
	mask[i] = 1;
    }
/* Estimate the full model: */
    vmmin(*bdim, beta, &Fmin,
	  we_fun, gwe_fun, maxiter, trace,  
	  mask, *eps, *eps, nREPORT,
	  ex, &fncount, &grcount, fail);

    if (trace)
      Rprintf("\nAfter 'vmmin': loglik = %f\n", -Fmin);
    loglik[1] = -Fmin;

    gwe_fun(*bdim, beta, dloglik, ex);
    
    if (trace){
      Rprintf("\nEfter 'vmmin': loglik = %f\n", -Fmin);
      Rprintf(" beta och dloglik:\n");
      for (i = 0; i < *bdim; i++){
	Rprintf("%f, %f\n", beta[i], dloglik[i]);
      }
    }

/*
    ord = 1;

    F77_CALL(wfunc)(&ord, &ipfixed, ex->pfix, bdim, ex->mb, beta,
		    ex->nn, ex->z, ex->time0, ex->time, ex->ind, ex->offset,
		    ex->f, ex->fp, ex->fpp, ex->iok);
*/


/*
    ord = 2;

    sw_fun(&ord, 
	   bdim, ncov, beta, 
	   nn, covar, time0, time, ind, 
	   offset, ns, nstra,
	   &Fmin, dloglik, variance, fail);

    loglik[1] = -Fmin;
*/

/* Take some Newton-Raphson steps: */
    F77_CALL(weibnr)(iter, eps, printlevel, nn, ncov, bdim,
		     time0, time, ind, covar, offset,
		     beta, loglik + 1, dloglik,
		     variance, ns, nstra,
		     conver, fail);
    if (trace){
      Rprintf("Variance (in [weibreg]) after N-R:\n");
      for (i = 0; i < *bdim; i++){
	for (j = 0; j < *bdim; j++){
	  Rprintf("%f ", variance[i + j * (*bdim)]);
	}
	Rprintf("\n");
      }

    Rprintf("Score: ");
    for (i = 0; i < *bdim; i++) Rprintf("%f ", dloglik[i]);
    Rprintf("\n");
    }
    if (trace){
	Rprintf("\nAfter Newton-Raphson: loglik = %f\n", loglik[1]);
	Rprintf("fail = %d\n", *fail);
    }
 }

File: R/weibreg.R
#' Weibull Regression
#' 
#' Proportional hazards model with baseline hazard(s) from the Weibull family
#' of distributions.  Allows for stratification with different scale and shape
#' in each stratum, and left truncated and right censored data.
#' 
#' The parameterization is the same as in \code{\link{coxreg}} and
#' \code{\link[survival]{coxph}}, but different from the one used by
#' \code{\link[survival]{survreg}}. The model is \deqn{h(t; a, b, \beta, z) =
#' (a/b) (t/b)^{a-1} exp(z\beta)}{% h(t; a, b, beta, z) = (a/b) (t/b)^(a-1)
#' exp(z beta)} This is in correspondence with \code{\link{Weibull}}. To
#' compare regression coefficients with those from \code{survreg} you need to
#' divide by estimated shape (\eqn{\hat{a}}{a}) and change sign. The p-values
#' and test statistics are however the same, with one exception; the score test
#' is done at maximized scale and shape in \code{weibreg}.
#' 
#' This model is a Weibull distribution with shape parameter \eqn{a} and scale
#' parameter \eqn{b \exp(-z\beta / a)}{b exp(-z beta / a)}
#' 
#' @param formula a formula object, with the response on the left of a ~
#' operator, and the terms on the right.  The response must be a survival
#' object as returned by the Surv function.
#' @param data a data.frame in which to interpret the variables named in the
#' formula.
#' @param na.action a missing-data filter function, applied to the model.frame,
#' after any subset argument has been used.  Default is
#' \code{options()$na.action}.
#' @param init vector of initial values of the iteration.  Default initial
#' value is zero for all variables.
#' @param shape If positive, the shape parameter is fixed at that value (in
#' each stratum).  If zero or negative, the shape parameter is estimated.  If
#' more than one stratum is present in data, each stratum gets its own
#' estimate.
#' @param control a list with components \code{eps} (convergence criterion),
#' \code{maxiter} (maximum number of iterations), and \code{silent} (logical,
#' controlling amount of output).  You can change any component without mention
#' the other(s).
#' @param singular.ok Not used.
#' @param model Not used.
#' @param x Return the design matrix in the model object?
#' @param y Return the response in the model object?
#' @param center Deprecated, and not used.  Will be removed in the future.
#' @return A list of class \code{c("weibreg", "coxreg")} with components
#' \item{coefficients}{Fitted parameter estimates.} \item{var}{Covariance
#' matrix of the estimates.} \item{loglik}{Vector of length two; first
#' component is the value at the initial parameter values, the second componet
#' is the maximized value.} \item{score}{The score test statistic (at the
#' initial value).} \item{linear.predictors}{The estimated linear predictors.}
#' \item{means}{Means of the columns of the design matrix.}
#' \item{w.means}{Weighted (against exposure time) means of covariates;
#' weighted relative frequencies of levels of factors.} \item{n}{Number of
#' spells in indata (possibly after removal of cases with NA's).}
#' \item{events}{Number of events in data.} \item{terms}{Used by extractor
#' functions.} \item{assign}{Used by extractor functions.} %
#' \item{wald.test}{The Wald test statistic (at the initial value).}
#' \item{y}{The Surv vector.} \item{isF}{Logical vector indicating the
#' covariates that are factors.} \item{covars}{The covariates.}
#' \item{ttr}{Total Time at Risk.} \item{levels}{List of levels of factors.}
#' \item{formula}{The calling formula.} \item{call}{The call.}
#' \item{method}{The method.} \item{convergence}{Did the optimization
#' converge?} \item{fail}{Did the optimization fail? (Is \code{NULL} if not).}
#' \item{pfixed}{TRUE if shape was fixed in the estimation.}
#' @note This function is not maintained, and may behave in unpredictable ways.
#' Use \code{\link{phreg}} with \code{dist = "weibull"} (the default) instead!
#' Will soon be declared deprecated.
#' @section Warning: The print method \code{\link{print.weibreg}} doesn't work
#' if threeway or higher order interactions are present.
#' 
#' Note further that covariates are internally centered, if \code{center =
#' TRUE}, by this function, and this is not corrected for in the output. This
#' affects the estimate of \eqn{\log(scale)}{log(scale)}, but nothing else. If
#' you don't like this, set \code{center = FALSE}.
#' @author Göran Broström
#' @seealso \code{\link{phreg}}, \code{\link{coxreg}},
#' \code{\link{print.weibreg}}
#' @keywords survival regression
#' @examples
#' 
#'  dat <- data.frame(time = c(4, 3, 1, 1, 2, 2, 3),
#'                 status = c(1, 1, 1, 0, 1, 1, 0),
#'                 x = c(0, 2, 1, 1, 1, 0, 0),
#'                 sex = c(0, 0, 0, 0, 1, 1, 1))
#'  weibreg( Surv(time, status) ~ x + strata(sex), data = dat) #stratified model
#' 
#' @export weibreg
weibreg <-
  function (formula = formula(data),
            data = parent.frame(), 
            na.action = getOption("na.action"),
            init,
            shape = 0, ## Means shape is estimated, ie true Weibull; > 0 fixed!
            control = list(eps = 1e-4, maxiter = 10, trace = FALSE),
            singular.ok = TRUE,
            model = FALSE, 
            x = FALSE,
            y = TRUE,
            center = TRUE) 
{
    
    pfixed <- (shape > 0)
    call <- match.call()
    m <- match.call(expand.dots = FALSE)
    temp <- c("", "formula", "data", "na.action")
    m <- m[match(temp, names(m), nomatch = 0)]
    
    special <- "strata"
    Terms <- if (missing(data)) 
      terms(formula, special)
    else terms(formula, special, data = data)
    m$formula <- Terms
    m[[1]] <- as.name("model.frame")
    m <- eval(m, parent.frame())
    
    Y <- model.extract(m, "response")
    if (!inherits(Y, "Surv")) 
      stop("Response must be a survival object")
    
    weights <- model.extract(m, "weights")
    offset <- attr(Terms, "offset")
    tt <- length(offset)
    offset <- if (tt == 0) 
      rep(0, nrow(Y))
    else if (tt == 1) 
      m[[offset]]
    else {
        ff <- m[[offset[1]]]
        for (i in 2:tt) ff <- ff + m[[offset[i]]]
        ff
    }
    attr(Terms, "intercept") <- 1
    strats <- attr(Terms, "specials")$strata
    dropx <- NULL
    
    if (length(strats)) {
        temp <- survival::untangle.specials(Terms, "strata", 1)
        dropx <- c(dropx, temp$terms)
        if (length(temp$vars) == 1) 
          strata.keep <- m[[temp$vars]]
        else strata.keep <- survival::strata(m[, temp$vars], shortlabel = TRUE)
        strats <- as.numeric(strata.keep)
    }
    if (length(dropx)) 
      newTerms <- Terms[-dropx]
    else newTerms <- Terms
    X <- model.matrix(newTerms, m)
    assign <- lapply(attrassign(X, newTerms)[-1], function(x) x - 1)
    X <- X[, -1, drop = FALSE]
    ncov <- NCOL(X)
    #########################################

    if (ncov){
        if (length(dropx)){
            covars <- names(m)[-c(1, (dropx + 1))]
        }else{
            covars <- names(m)[-1]
        }
        
        isF <- logical(length(covars))
        for (i in 1:length(covars)){
            if (length(dropx)){
                isF[i] <- ( is.factor(m[, -(dropx + 1)][, (i + 1)]) ||
                           is.logical(m[, -(dropx + 1)][, (i + 1)]) )
            }else{
                isF[i] <- ( is.factor(m[, (i + 1)]) ||
                           is.logical(m[, (i + 1)]) )
            }      
        }
        
        if (ant.fak <- sum(isF)){
            levels <- list()
            index <- 0
            for ( i in 1:length(covars) ){
                if (isF[i]){
                    index <- index + 1
                    if (length(dropx)){
                        levels[[i]] <- levels(m[, -(dropx + 1)][, (i + 1)])
                    }else{
                        levels[[i]] <- levels(m[, (i + 1)])
                    }
                }else{
                    ##cat("NULL level no  ", i, "\n")
                    levels[[i]] <- NULL
                }
            }
        }else{
            levels <- NULL
        }
    }
    
##########################################
    type <- attr(Y, "type")
    if (type != "right" && type != "counting") 
      stop(paste("This model doesn't support \"", type, "\" survival data", 
                 sep = ""))
    
    if (NCOL(Y) == 2){
        Y <- cbind(numeric(NROW(Y)), Y)
    }
    
    n.events <- sum(Y[, 3] != 0)
    if (n.events == 0) stop("No events; no sense in continuing!")
    if (missing(init)) 
      init <- NULL
    
    if (is.list(control)){
        if (is.null(control$eps)) control$eps <- 1e-4
        if (is.null(control$maxiter)) control$maxiter <- 10
        if (is.null(control$trace)) control$trace <- FALSE
    }else{
        stop("control must be a list")
    }
    
    
    fit <- weibreg.fit(X, 
                       Y,
                       strats,
                       offset,
                       init,
                       shape,
                       control,
                       center)

    if (ncov){
        fit$linear.predictors <- offset + X %*% fit$coefficients[1:ncov]
        fit$means <- apply(X, 2, mean)
    }else{
        fit$linear.predictors <- NULL
        fit$means <- NULL
    }
    ##score <- exp(lp)


    if (!fit$fail){
        fit$fail <- NULL
    }else{
        out <- paste("Singular hessian; suspicious variable No. ",
                     as.character(fit$fail), ":\n",
                     names(coefficients)[fit$fail], " = ",
                     as.character(fit$value),
                     "\nTry running with fixed shape", sep = "")
        stop(out)
    }
    
    
    fit$convergence <- as.logical(fit$conver)
    fit$conver <- NULL ## Ugly!
    
###########################################################################    
    ## Crap dealt with ......
    
    if (is.character(fit)) {
        fit <- list(fail = fit)
        class(fit) <- "mlreg"
    }
    else if (is.null(fit$fail)){
        if (!is.null(fit$coef) && any(is.na(fit$coef))) {
            vars <- (1:length(fit$coef))[is.na(fit$coef)]
            msg <- paste("X matrix deemed to be singular; variable", 
                         paste(vars, collapse = " "))
            if (singular.ok) 
              warning(msg)
            else stop(msg)
        }
        fit$n <- nrow(Y)
        fit$terms <- Terms
        fit$assign <- assign
        if (FALSE){ ## Out-commented...(why?)
            if (length(fit$coef) && is.null(fit$wald.test)) {
                nabeta <- !is.na(fit$coef)
                if (is.null(init)) 
                  temp <- fit$coef[nabeta]
                else temp <- (fit$coef - init)[nabeta]
                ##fit$wald.test <-
                  ##survival:::coxph.wtest(fit$var[nabeta, nabeta],   
                    ##                     temp, control$toler.chol)$test
            }
        }
        na.action <- attr(m, "na.action")
        if (length(na.action)) 
          fit$na.action <- na.action
        if (model) 
          fit$model <- m
        if (x) {
            fit$x <- X
            if (length(strats)) 
              fit$strata <- strata.keep
        }
        if (y) 
          fit$y <- Y
    }
    ##if (!is.null(weights) && any(weights != 1)) 
    ##    fit$weights <- weights
    
##########################################
    s.wght <- (Y[, 2] - Y[, 1])## * weights
    fit$ttr <- sum(s.wght)
    if (ncov){
        fit$isF <- isF
        fit$covars <- covars
        fit$w.means <- list()
        for (i in 1:length(fit$covars)){
            nam <- fit$covars[i]
            col.m <- which(nam == names(m))
            if (isF[i]){
                n.lev <- length(levels[[i]])
                fit$w.means[[i]] <- numeric(n.lev)
                for (j in 1:n.lev){
                    who <- m[, col.m] == levels[[i]][j]
                    fit$w.means[[i]][j] <-
                      sum( s.wght[who] ) / fit$ttr ## * 100, if in per cent
                }
            }else{
                fit$w.means[[i]] <- sum(s.wght * m[, col.m]) / fit$ttr
            }
        }
        fit$means <- colMeans(X)
        
    }
    
##########################################
    fit$ttr <- sum(s.wght)
    ##names(fit$coefficients) <- coef.names 
    fit$levels <- levels
    fit$formula <- formula(Terms)
    fit$call <- call
    fit$n.events <- n.events 
    ##class(fit) <- c("weibreg", "coxreg", "coxph")
    class(fit) <- c("weibreg", "phreg")
    fit$pfixed <- pfixed
    fit
}
